File: scala/Practice2/DynamicProgramming/Bit/TravelingSalesmanProblem.scala
// http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=DPL_2_A

object Main {

  implicit class Pipeline[T](x: T) {
    def |>[S](f: T => S): S = f(x)
  }

  def main(args: Array[String]) = (
    (() => io.StdIn.readLine().split(' ').map(_.toInt))
    |> (readLine => (
      readLine()
      |> (arr => (arr(0), arr(1)))
      |> { case (v, e) => (
        (
          (1 to e)
            .map(_x => (readLine() |> (arr => (arr(0), arr(1), arr(2)))))
            .map { case (s, t, d) => (s, (if (t == 0) v else t), d) },
          Array.range(0, 1 << v + 1).map(_x => Array.fill[Int](v + 1)(Int.MaxValue))
        )
        |> { case (graph, memo) => (
          (memo(1)(0) = 0)
          |> { case () =>
            (
              for (
                vx <- 1 to 1 << v + 1;
                (s, t, d) <- graph
                if (vx & 1) == 1 && (vx >> s & 1) == 1 && (vx >> t & 1) == 1
              )
              yield (vx, s, t, d)
            )
            .foreach { case (vx, s, t, d) => (
              memo(vx ^ (1 << t))(s)
              |> (prev =>
                if (prev != Int.MaxValue)
                  memo(vx)(t) = (memo(vx)(t) min (prev + d))
                else
                  ()
              )
            )}
          }
          |> (_x => memo.last.last)
        )}
      )}
    ))
    |> (x => if (x == Int.MaxValue) -1 else x)
    |> println
  )
}

// http://judge.u-aizu.ac.jp/onlinejudge/review.jsp?rid=5245114

File: fs/Practice2/DynamicProgramming/Bit/TravelingSalesmanProblem.fs
// http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=DPL_2_A

(fun () -> stdin.ReadLine () |> fun x -> x.Split ' ' |> Array.map int)
|> fun readLine ->
  readLine ()
  |> fun arr -> (arr.[0], arr.[1])
  |> fun (v, e) ->
    (
      List.init e (fun _ -> readLine () |> fun arr -> (arr.[0], arr.[1], arr.[2]))
      |> List.map (fun (s, t, d) -> (s, (if t = 0 then v else t), d)),
      Array.init (1 <<< (v + 1)) (fun _ -> Array.create (v + 1) System.Int32.MaxValue)
    )
    |> fun (graph, memo) ->
      (memo.[1].[0] <- 0)
      |> fun () ->
        seq {
          for vx in 1..1 <<< (v + 1) do 
          for (s, t, d) in graph do
          if vx &&& 1 = 1
            && vx >>> s &&& 1 = 1
            && vx >>> t &&& 1 = 1
          then yield (vx, t, d, memo.[vx ^^^ (1 <<< t)].[s])
        }
        |> Seq.filter (
          fun (_, _, _, prev) ->
            prev <> System.Int32.MaxValue
        )
        |> Seq.iter (
          fun (vx, t, d, prev) ->
            memo.[vx].[t] <- min memo.[vx].[t] (prev + d)
        )
      |> fun _ -> memo |> Array.last |> Array.last
      |> fun x -> if x = System.Int32.MaxValue then -1 else x
|> printfn "%d"

// input:
// 4 6
// 0 1 2
// 1 2 3
// 1 3 9
// 2 0 1
// 2 3 6
// 3 2 4

// memo:
// [|
//   [| -;  -;  -;  -;  - |]
//   [| 0;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  2;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  5;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -; 11;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -; 15; 11;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  6 |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -;  - |]
//   [| -;  -;  -;  -; 16 |]
// |]

// output:
// 16
